lib/init/grass.py

```diff
@@ -2555,39 +2509,87 @@ def main():
+                gisrc,
+                os.path.join(
+                    mapset_settings.gisdbase,
+                    mapset_settings.location,
+                    mapset_settings.mapset,
+                ),
             )
-            debug(f"demo_mapset_usable: {demo_mapset_usable}")
-            if demo_mapset_usable:
-                set_mapset_to_gisrc(
-                    gisrc=gisrc, grassdb=grassdb, location=location, mapset=mapset
+            if is_first_time_user():
+                # Ensure default data hiearchy
+                (
+                    default_gisdbase,
+                    default_location,
+                    default_mapset,
+                ) = ga.ensure_default_data_hierarchy()
+
+                if default_mapset is None:
+                    sys.exit("Failed to start GUI, GRASS GIS is not running.")
+
+                default_mapset_path = os.path.join(
+                    default_gisdbase,
+                    default_location,
+                    default_mapset,
                 )
-            else:
-                # Try interactive startup
-                # User selects LOCATION and MAPSET if not set
-                if not set_mapset_interactive(grass_gui):
-                    # No GUI available, update gisrc file
-                    fatal(
-                        _(
-                            "<{0}> requested, but not available. Run GRASS in text "
-                            "mode (--text) or install missing package (usually "
-                            "'grass-gui')."
-                        ).format(grass_gui)
+                # First session
+                if can_start_in_mapset(
+                    mapset_path=default_mapset_path, ignore_lock=False
+                ):
+                    # Write mapset info to gisrc file
+                    add_mapset_to_gisrc(
+                        gisrc=gisrc,
+                        grassdb=default_gisdbase,
+                        location=default_location,
+                        mapset=default_mapset,
+                    )
+                # Another first-time session
+                else:
+                    # Add last used mapset to gisrc
+                    add_last_mapset_to_gisrc(gisrc, default_mapset_path)
+
+                    # Create temporary location
+                    params.tmp_location = True
+                    set_mapset(
+                        gisrc=gisrc,
+                        geofile="XY",
+                        create_new=True,
+                        tmp_location=params.tmp_location,
+                        tmpdir=tmpdir,
                     )
+            else:
+                # Create temporary location
+                params.tmp_location = True
+                set_mapset(
+                    gisrc=gisrc,
+                    geofile="XY",
+                    create_new=True,
+                    tmp_location=params.tmp_location,
+                    tmpdir=tmpdir,
+                )
```

Comment:
You define the XY part in two places. Ideally restructure this code: 1) The ifs are too long to navigate. 2) The whole "not-usable" block adds too many lines to main(). What about a separate function?

Reply:
Well, I know that it is hard to navigate.. but there is quite a lot of cases. We have first-time user, then first-time user with fallback session (e.g. second startup), and then normal fallback session...
I tried to simplify it a bit and defined a new function `create_fallback_session`. But I am afraid that "ifs" cannot be more simplified than I did it now. Maybe you will have better idea how to do it.